### PR TITLE
Get accesskey from service

### PR DESCRIPTION
**Pull request by terencefan**

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 1]
```diff
@@ -1,18 +1,82 @@
...
+        // public DateTime ExpireAt = DateTime.UtcNow;
+
+        public AccessKey(string key = null)
         {
             Value = key;
-            Id = key.GetHashCode().ToString();
+
+            if (!string.IsNullOrEmpty(key))
+            {
+                Id = key.GetHashCode().ToString();
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:31]: `string.GetHashCode()` may not work when cross process.

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 2]
```diff
@@ -41,7 +43,16 @@ public ServiceEndpoint(string connectionString, EndpointType type = EndpointType
             {
                 string key;
                 (Endpoint, key, Version, Port) = ConnectionStringParser.Parse(connectionString);
-                AccessKey = new AccessKey(key);
+
+                if (string.IsNullOrEmpty(key))
+                {
+                    AccessKey = new AccessKey();
+                    _ = AccessKey.AuthorizeTask(Endpoint, Port ?? 80);
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:50]: `80`? should it be `443`?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 3]
```diff
@@ -1,18 +1,82 @@
...
+        {
+            if (AuthorizedTask.IsCompleted)
+            {
+                return;
+            }
+
+            var span = TimeSpan.FromMinutes(30);
+            var timer = new TimerAwaitable(span, span);
+
+            while (await timer)
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:75]: no quit?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 4]
```diff
@@ -1,18 +1,82 @@
...
+            if (!string.IsNullOrEmpty(key))
+            {
+                Id = key.GetHashCode().ToString();
+                _authorizedTcs.SetResult(true);
+            }
+        }
+
+        private async Task Authorize(string endpoint, int port, string accessToken)
+        {
+            var client = new HttpClient();
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:38]: use `IHttpClientFactory` or dispose it.

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 5]
```diff
@@ -1,18 +1,82 @@
...
+
+            var builder = new UriBuilder(endpoint + "/accessKey")
+            {
+                Port = port,
+                Query = query.ToString()
+            };
+            HttpResponseMessage response = await client.GetAsync(builder.ToString());
+
+            var json = await response.Content.ReadAsStringAsync();
+            var obj = JsonConvert.DeserializeObject<IDictionary<string, string>>(json);
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:52]: use `JObject`?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs [hunk 6]
```diff
@@ -13,10 +13,12 @@ public class ServiceEndpoint
 
         public virtual string Name { get; internal set; }
 
+        internal bool _online = true;
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/ServiceEndpoint.cs:16]: use property like `OnlineInternal`?

File: src/Microsoft.Azure.SignalR.AspNet/ServiceOptions.cs [hunk 7]
```diff
@@ -72,6 +72,8 @@ public class ServiceOptions : IServiceEndpointOptions
         /// </summary>
         public int? DisconnectTimeoutInSeconds { get; set; }
 
+        public AzureAdOptions AzureAdOptions { get; set; } = new AzureAdOptions();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/ServiceOptions.cs:75]: Add an auth option and move aad options inside, in case we can support more auth type later?

File: src/Microsoft.Azure.SignalR.AspNet/EndpointProvider/ServiceEndpointProvider.cs [hunk 8]
```diff
@@ -46,8 +47,13 @@ public ServiceEndpointProvider(ServiceEndpoint endpoint, ServiceOptions options)
             _port = endpoint.Port;
             _algorithm = options.AccessTokenAlgorithm;
             Proxy = options.Proxy;
-        }
 
+            if (!_accessKey.InitializedTask.IsCompleted)
+            {
+                var app = AADHelper.BuildApplication(options.AzureAdOptions);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/EndpointProvider/ServiceEndpointProvider.cs:53]: `AAD`=> let's stick to one naming convention for aad all the way through: `AzureActiveDirectory` or `Aad`

File: src/Microsoft.Azure.SignalR.AspNet/EndpointProvider/ServiceEndpointProvider.cs [hunk 9]
```diff
@@ -1,6 +1,7 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Microsoft.Azure.SignalR.Common.Utilities;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/EndpointProvider/ServiceEndpointProvider.cs:4]: system namespaces first

File: src/Microsoft.Azure.SignalR.AspNet/EndpointProvider/ServiceEndpointProvider.cs [hunk 10]
```diff
@@ -46,8 +47,13 @@ public ServiceEndpointProvider(ServiceEndpoint endpoint, ServiceOptions options)
             _port = endpoint.Port;
             _algorithm = options.AccessTokenAlgorithm;
             Proxy = options.Proxy;
-        }
 
+            if (!_accessKey.InitializedTask.IsCompleted)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.AspNet/EndpointProvider/ServiceEndpointProvider.cs:51]: Checking `IsCompleted` sounds a bit tricky to me. we can get if AAD auth is necessary from options?

File: src/Microsoft.Azure.SignalR.AspNet/ServiceOptions.cs (hunk 7 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.AspNet/ServiceOptions.cs:75]: make it internal first

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        public readonly string ClientId;
+        public readonly string ClientSecret;
+        public readonly X509Certificate2 ClientCert;
+
+        public readonly string TenantId;
+
+        public string Instance = "https://login.microsoftonline.com/{0}";
+
+        public AzureAdOptions()
+        {
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs:20]: having a default ctor with `Enabled` as false looks a bit strange

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
+﻿using System.Security.Cryptography.X509Certificates;
+
+namespace Microsoft.Azure.SignalR
+{
+    public class AzureAdOptions
+    {
+        public bool Enabled = false;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs]: no public fields

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs]: I think we can use a similar format as MADL client lib https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/src/client/Microsoft.Identity.Client/AppConfig/ApplicationOptions.cs

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 13]
```diff
@@ -1,18 +1,74 @@
...
+
+        private volatile string _kid;
+
+        public Task<bool> InitializedTask => _initializedTcs.Task;
+
+        public bool Initialized => InitializedTask.IsCompleted && InitializedTask.Result;
+
+        public AccessKey(string key = null)
+        {
+
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:29]: remove extra empty line

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 14]
```diff
@@ -1,18 +1,74 @@
...
+            if (!string.IsNullOrEmpty(key))
+            {
+                _accessKey = key;
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:32]: what happens when aad not enabled while key is null?

When key is not null, what if msi is enabled? can it happen?

File: AzureSignalR.sln [hunk 15]
```diff
@@ -27,6 +27,9 @@ EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Azure.SignalR.Protocols", "src\Microsoft.Azure.SignalR.Protocols\Microsoft.Azure.SignalR.Protocols.csproj", "{2EBFCD26-6E31-4855-ABF5-0A82E78F87C7}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{2429FBD8-1FCE-4C42-AA28-DF32F7249E77}"
+	ProjectSection(SolutionItems) = preProject
```
Comment (vicancy) [inline AzureSignalR.sln:30]: what is this?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 16]
```diff
@@ -1,18 +1,74 @@
...
+            var scopes = new string[] { ".default" };
+            var token = await app.AcquireTokenForClient(scopes).WithSendX5C(true).ExecuteAsync();
+            await key.Authorize(endpoint, port, token.AccessToken);
+        }
+
+        private async Task Authorize(string endpoint, int? port, string accessToken)
         {
-            Value = key;
-            Id = key.GetHashCode().ToString();
+            using var client = new HttpClient();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:47]: never new HttpClient...

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 17]
```diff
@@ -1,18 +1,74 @@
...
+            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
+
+            var query = HttpUtility.ParseQueryString(string.Empty);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:50]: what's the meaning of parsing a string.Empty?

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs [hunk 18]
```diff
@@ -1,18 +1,74 @@
...
+            query.Add("serverId", "localhost");
+
+            var builder = new UriBuilder(endpoint + "/auth/accessKey")
+            {
+                Port = port ?? 443,
+                Query = query.ToString()
+            };
+            HttpResponseMessage response = await client.GetAsync(builder.ToString());
+
+            var json = await response.Content.ReadAsStringAsync();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:60]: There is a https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR.Management/RestClient.cs , consider reuse

File: src/Microsoft.Azure.SignalR.Common/Utilities/AADHelper.cs [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+                }
+                else
+                {
+                    builder.WithClientSecret(options.ClientSecret);
+                }
+                return builder.Build();
+            }
+            else
+            {
+                throw new System.Exception("Failed to build Azure AD Application. (AAD disabled)");
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Utilities/AADHelper.cs:27]: never throw general `Exception`

File: src/Microsoft.Azure.SignalR.Common/Utilities/ConnectionStringParser.cs [hunk 20]
```diff
@@ -50,7 +50,7 @@ internal static (string endpoint, string accessKey, string version, int? port) P
                 dict.Add(key, kvp[1].Trim());
             }
 
-            if (!dict.ContainsKey(EndpointProperty) || !dict.ContainsKey(AccessKeyProperty))
+            if (!dict.ContainsKey(EndpointProperty))
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Utilities/ConnectionStringParser.cs:53]: is this a breaking change? What happens when user does not enable Aad but missing `accesskey`?

File: src/Microsoft.Azure.SignalR/HubHost/NegotiateHandler.cs [hunk 21]
```diff
@@ -51,7 +51,10 @@ public async Task<NegotiationResponse> Process(HttpContext context, string hubNa
             var request = context.Request;
             var cultureName = context.Features.Get<IRequestCultureFeature>()?.RequestCulture.Culture.Name;
             var originalPath = GetOriginalPath(request.Path);
-            var provider = _endpointManager.GetEndpointProvider(_router.GetNegotiateEndpoint(context, _endpointManager.GetEndpoints(hubName)));
+
+            var endpoints = _endpointManager.GetEndpoints(hubName);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/HubHost/NegotiateHandler.cs:55]: revert the change.

The rule is to avoid including a file if only refactor changes are involved in this file. The refactor changes can be done in a separate PR.

File: src/Microsoft.Azure.SignalR/HubHost/ServerLifetimeManager.cs [hunk 22]
```diff
@@ -20,7 +20,6 @@ IServiceProvider provider
 #elif NETSTANDARD
             var lifetime = provider.GetService<IApplicationLifetime>();
 #endif
-
             lifetime?.ApplicationStopping.Register(Shutdown);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/HubHost/ServerLifetimeManager.cs:23]: same here, revert the change

File: src/Microsoft.Azure.SignalR/ServiceOptions.cs [hunk 23]
```diff
@@ -59,6 +59,11 @@ public class ServiceOptions : IServiceEndpointOptions
         /// </summary>
         public ServerStickyMode ServerStickyMode { get; set; } = ServerStickyMode.Disabled;
 
+        /// <summary>
+        /// 
+        /// </summary>
+        public AzureAdOptions AzureAdOptions { get; set; } = new AzureAdOptions();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR/ServiceOptions.cs:65]: internal first if not separated to another PR

File: AzureSignalR.sln (hunk 15 above)
Comment (terencefan) [reply AzureSignalR.sln:30]: IDK, just removed it.

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs (hunk 12 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs]: Good idea.

File: src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs (hunk 17 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Endpoints/AccessKey.cs:50]: To initialize a Query object.

File: src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs (hunk 11 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs:20]: How about we make the default ctor private and expose a static `Disabled` instance?

Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Auth/AzureADOptions.cs:20]:
```C#
        public static AzureAdOptions Disabled = new AzureAdOptions();

        private AzureAdOptions()
        {
        }
```
